tests/future/basic.rs

```diff
@@ -221,6 +221,78 @@ fn async_counter() {
     });
 }
 
+// We need a way to hold the `MutexGuard`, which is `!Send`, across an `await`.
+struct WrappedMutexGuard<'a> {
+    #[allow(unused)]
+    guard: shuttle::sync::MutexGuard<'a, ()>,
+}
+
+unsafe impl<'a> Send for WrappedMutexGuard<'a> {}
+
+async fn acquire_and_loop(mutex: Arc<Mutex<()>>) {
+    let _g = WrappedMutexGuard {
+        guard: mutex.lock().unwrap(),
+    };
+    loop {
+        future::yield_now().await;
+    }
+}
+
+// The idea is to acquire a mutex, abort the JoinHandle, then acquire the Mutex.
+// This should succeed, because `JoinHandle::abort()` should free the Mutex.
+#[test]
+fn abort_frees_mutex() {
+    check_random(
+        || {
+            let mutex = Arc::new(Mutex::new(()));
+            let jh = future::spawn(acquire_and_loop(mutex.clone()));
+
+            jh.abort(); // this unblocks
+
+            let _g = mutex.lock();
+        },
+        1000,
+    );
+}
+
+// The idea is to acquire a mutex, drop the JoinHandle, then acquire the Mutex.
+// This should fail, because `drop`ping the JoinHandle just detaches it, meaning
+// it keeps holding the Mutex.
+#[test]
+#[should_panic(expected = "exceeded max_steps bound")]
+fn drop_join_handle_deadlocks() {
+    check_random(
+        || {
+            let mutex = Arc::new(Mutex::new(()));
+            let jh = future::spawn(acquire_and_loop(mutex.clone()));
+
+            drop(jh);
+
+            let _g = mutex.lock();
+        },
+        1000,
+    );
+}
+
+// The idea is to acquire a mutex, forget the JoinHandle, then acquire the Mutex.
+// This should fail, because `forget`ting the JoinHandle doesn't cause it to release
+// the Mutex.
+#[test]
+#[should_panic(expected = "exceeded max_steps bound")]
+fn forget_join_handle_deadlocks() {
+    check_random(
+        || {
+            let mutex = Arc::new(Mutex::new(()));
+            let jh = future::spawn(acquire_and_loop(mutex.clone()));
+
+            std::mem::forget(jh);
+
+            let _g = mutex.lock();
+        },
+        1000,
+    );
+}
+
```

Comment:
Something like this doesn't work, because it's possible for the abort to happen while `t2` is blocked acquiring the lock:
```rust
#[test]
fn abort_while_blocked() {
    check_dfs(
        || {
            let lock = Arc::new(Mutex::new(0i32));
            let lock2 = lock.clone();
            let t1 = thread::spawn(move || {
                let _l = lock2.lock().unwrap();
            });
            let t2 = future::spawn(async move {
                let _l = lock.lock().unwrap();
            });
            t2.abort();
            t1.join().unwrap();
        },
        None,
    );
}
```
The way it fails right now is a bit funny — when `t1` drops the lock, it [asserts that all the waiters are still blocked](https://github.com/awslabs/shuttle/blob/9076a444c7f611eeb94fa0b562263bf29a178fbd/src/sync/mutex.rs#L277-L278), but when the future aborts it's no longer blocked.

Not sure what the right thing to do here is. Maybe we need to check lazily whether a task has been aborted, only once [a poll of the `Wrapper` task returns Pending](https://github.com/awslabs/shuttle/blob/9076a444c7f611eeb94fa0b562263bf29a178fbd/src/future/mod.rs#L173)? But that might be more trouble than it's worth for what you want to do.

Reply:
Hmm, what about removing the `assert!` ([which is repeated in `unblock` anyways](https://github.com/awslabs/shuttle/blob/9076a444c7f611eeb94fa0b562263bf29a178fbd/src/runtime/task/mod.rs#L363)), and then updating `unblock` to allow state finished as well ? I need to do a pass and check that we don't ever do "silent passing" of synchro primitives if so. If we do "silent passing" then the "passer" has to check that the task it passes to isn't finished.

---

Actually, wait, is our implementation of `MutexGuard::drop` correct? Aren't `std::sync::Mutex`es fair?

Reply:
Their fairness is platform-dependent IIRC.

Removing the assert might be OK, though we should add an explicit aborted flag as a sanity check that the only way a task gets unblocked is if it was aborted. But tokio's JoinHandle semantics are that the task cancels at the next `await` point. I worry about aborting a task mid-synchronization operation because it leaves data structures in an intermediate state that isn't reachable in real code. Like this:
```rust
let t1 = future::spawn(async move {
    let l1 = lock1.lock().unwrap();
    *l1 += 1;
    let l2 = lock2.lock().unwrap();
    *l2 += 1;
});
t1.abort();
assert_eq!(*l1.lock().unwrap(), *l2.lock().unwrap());
```
In real Rust this assertion can't fail because the task body is atomic from the perspective of `abort`, but in our implementation it can.

Reply:
Hmm yeah, the immediate `drop` of the `Continuation` doesn't seem to be the correct solution. `drop`ping immediately is correct if it has never been scheduled, else we need to do it at the next `await` point. I guess lazily checking in `Wrapper` is the correct solution.